profiler/internal/cmemprof/pprof.go

```diff
@@ -6,13 +6,132 @@
 package cmemprof
 
 import (
+	"bytes"
 	"os"
 	"runtime"
+	"strconv"
 	"strings"
 
 	"github.com/google/pprof/profile"
 )
 
+var defaultMapping = &profile.Mapping{
+	ID:   1,
+	File: os.Args[0],
+}
+
+var mappings []*profile.Mapping
+
+func getMapping(addr uint64) *profile.Mapping {
+	for _, m := range mappings {
+		if addr >= m.Start && addr <= m.Limit {
+			return m
+		}
+	}
+	return defaultMapping
+}
+
+// bytes.Cut, but backported so we can still support Go 1.17
+func bytesCut(s, sep []byte) (before, after []byte, found bool) {
+	if i := bytes.Index(s, sep); i >= 0 {
+		return s[:i], s[i+len(sep):], true
+	}
+	return s, nil, false
+}
+
+func stringsCut(s, sep string) (before, after string, found bool) {
+	if i := strings.Index(s, sep); i >= 0 {
+		return s[:i], s[i+len(sep):], true
+	}
+	return s, "", false
+}
+
+func init() {
+	if runtime.GOOS != "linux" {
+		mappings = append(mappings, defaultMapping)
+		return
+	}
+
+	data, err := os.ReadFile("/proc/self/maps")
+	if err != nil {
+		mappings = append(mappings, defaultMapping)
+		return
+	}
+
+	// To have a more accurate profile, we need to provide mappings for the
+	// executable and linked libraries, since the profile might contain
+	// samples from outside the executable if the allocation profiler is
+	// used in conjunction with a cgo traceback library.
+	//
+	// /proc/self/maps parsing code lifted from runtime/pprof/proto.go, see
+	// https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/runtime/pprof/proto.go;l=596
```

Comment:
IANAL, but I think referencing the origin of the code is not enough. We need to include the [Go license](https://cs.opensource.google/go/go/+/master:LICENSE).

This could be done by adding it as a comment, perhaps at the bottom of this file, or as a separate file somewhere else and then referencing it in this comment.

Reply:
So I did a quick check on a Linux VM with a few basic things installed (something like `for pid in $(ps axh -o pid); do sudo grep 'r-xp' "/proc/${pid}/maps" | wc -l; done`) and I see a median of ~30 mappings with a high of 83 (for something called ModemManager). I did a little benchmark of looking for a random address with either a binary or linear search (assuming the mappings are sorted by start address) and at least on my machine you don't really break even for binary search until you pass around 30 mappings. On the other hand, even for fewer mappings the difference is only ~nanoseconds and we only do this search once per unique address. So I guess it doesn't hurt to just do a binary search.

Reply:
> it looks like this repo has a [third-party licenses file](https://github.com/DataDog/dd-trace-go/blob/main/LICENSE-3rdparty.csv) where we can track the code we're borrowing.

I don't think that file alone is sufficient. A copy of the actual license must be included. A reference to it is not enough. At least that's my understanding.

Reply:
Right. I'm going back to using the Go source code. To use the code, we need to:

1) Include the copyright notice
2) Include a copy of the license with the notice and disclaimer
3) For Datadog, include a reference to the license in LICENSE-3rdparty.csv